spec/consensus/proposer-based-timestamp/tla/TendermintPBT_002_draft.tla

```diff
@@ -831,53 +805,58 @@ ConsensusSafeValidCorrProp ==
         (/\ Proposer[pr] \in Corr
         /\ decision[p] = Decision(Proposal(v,t,pr), dr))
           \* then t is between the minimal and maximal initial local time
-          => /\ beginRound[pr] <= t 
+          => /\ beginRound[pr, p] <= t 
              /\ t <= lastBeginRound[pr]
 
-\* [PBTS-CONSENSUS-REALTIME-VALID-CORR.0]
-ConsensusRealTimeValidCorr ==
-  \A r \in Rounds :
-    \E p \in Corr, v \in ValidValues, t \in Timestamps, pr \in Rounds: 
-     (/\ decision[p] = Decision(Proposal(v,t,pr), r) 
-      /\ proposalTime[r] /= NilTimestamp)
-        => (/\ proposalTime[r] - Precision <= t
-            /\ t <= proposalTime[r] + Precision)
 
+
+\* @type: (PROPOSAL, ROUND) => Set(PREMESSAGE);
+POLSet(prop, rnd) == 
+  { msg \in msgsPrevote[rnd]: msg.id = prop}
+\* @type: (Set(PREMESSAGE)) => Bool;
+IsValidPOL(s) == Cardinality(s) >= THRESHOLD2
+
+\* @type: (Set(PREMESSAGE)) => Bool;
+ContainsPrevoteFromCorrect(set) ==
+  \E p \in Corr, msg \in set: msg.src = p
+
+\* [PBS-DERIVED-POL.1]
+DerivedProofOfLocks == 
+  \A r \in Rounds, prop \in ValidProposals:
+    LET t == prop[2] IN
+    LET PS == POLSet(prop, r) IN
+    ( 
+      /\ IsValidPOL(PS)
+      /\ ContainsPrevoteFromCorrect(PS)
+    ) =>
+    \E rStar \in Rounds:
```

Comment:
This looks good. But I would like to mention that with the definition of Proposal we are using here, this part could be simplified from the observation that `rStar = prop[3]`.

In other words, `r*` is stored in the Proposal, it is the first round in which a pair (value, time) was proposed.

Reply:
It's not immediately clear to me why this is the case. `prop[3]` is the round in which the proposal was made, whereas `rStar` tells us to look at `msgsPrevote[rStar]` in `POLSet`, i.e. it is the round when the prevote message was sent/received. As far as I understand, a proposal can be prevoted for in a later round than the one in which it was made.